singlepoll/linux.go

```diff
@@ -0,0 +1,194 @@
+// +build linux
+
+package singlepoll
+
+import (
+	"context"
+	"errors"
+	"runtime"
+	"sync"
+	"syscall"
+
+	"github.com/gxed/eventfd"
+	logging "github.com/ipfs/go-log"
+)
+
+var (
+	ErrUnsupportedMode error = errors.New("only 'w' and 'r' modes are supported on this arch")
+)
+
+var (
+	initOnce sync.Once
+	workChan chan interface{}
+	log      logging.EventLogger = logging.Logger("reuseport-poll")
+)
+
+type addPoll struct {
+	fd     int
+	events uint32
+	ctx    context.Context
+	wakeUp chan<- error
+}
+
+type ctxDone struct {
+	fd int
+}
+
+func PollPark(reqctx context.Context, fd int, mode string) error {
+	initOnce.Do(func() {
+		workChan = make(chan interface{}, 128)
+		go worker()
+	})
+
+	events := uint32(0)
+	for _, c := range mode {
+		switch c {
+		case 'w':
+			events |= syscall.EPOLLOUT
+		case 'r':
+			events |= syscall.EPOLLIN
+		default:
+			return ErrUnsupportedMode
+		}
+
+	}
+
+	wakeUp := make(chan error)
+	workChan <- addPoll{
+		fd:     fd,
+		events: events,
+		ctx:    reqctx,
+		wakeUp: wakeUp,
+	}
+
+	return <-wakeUp
+}
+
+func criticalError(msg string, err error) {
+	log.Errorf("%s: %s.", msg, err.Error())
+	log.Errorf("This is critical error, please report it at https://github.com/jbenet/go-reuseport/issues/new")
+	log.Errorf("Bailing out. You are on your own. Good luck.")
+
+	for {
+		select {
+		case <-backgroundctx.Done():
+			return
+		case unit := <-workChan:
+			switch u := unit.(type) {
+			case addPoll:
+				u.wakeUp <- err
+			default:
+			}
+		}
+	}
+}
+
+func worker() {
+	epfd, err := syscall.EpollCreate1(0)
+	if err != nil {
+		criticalError("EpollCreate1(0) failed", err)
+	}
+	evfd, err := eventfd.New()
+	if err != nil {
+		criticalError("eventfd.New() failed", err)
+	}
+
+	pool := make(map[int]addPoll)
+
+	{
+		event := syscall.EpollEvent{
+			Events: syscall.EPOLLIN,
+			Fd:     int32(evfd.Fd()),
+		}
+		syscall.EpollCtl(epfd, syscall.EPOLL_CTL_ADD, evfd.Fd(), &event)
+		runtime.KeepAlive(event) // won't hurt but might fix some possible issues with GC
```

Comment:
I miss understood use of the KeepAlive in this case, it is useful only if you wrap your fd in os.File